homeassistant/components/osoenergy/water_heater.py

```diff
@@ -0,0 +1,237 @@
+"""Support for OSO Energy water heaters."""
+from collections.abc import Mapping
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components.water_heater import (
+    WaterHeaterEntity,
+    WaterHeaterEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import UnitOfTemperature
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import config_validation as cv, entity_platform
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+import homeassistant.util.dt as dt_util
+
+from . import OSOEnergyEntity
+from .const import (
+    ATTR_UNTIL_TEMP_LIMIT,
+    ATTR_V40MIN,
+    DOMAIN,
+    EXTRA_HEATER_ATTR,
+    HEATER_MAX_TEMP,
+    HEATER_MIN_TEMP,
+    MANUFACTURER,
+    OPERATION_LIST,
+    OSO_ENERGY_TO_HASS_STATE,
+    SERVICE_SET_PROFILE,
+    SERVICE_SET_V40MIN,
+    SERVICE_TURN_OFF,
+    SERVICE_TURN_ON,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up OSO Energy heater based on a config entry."""
+    osoenergy = hass.data[DOMAIN][entry.entry_id]
+    devices = osoenergy.session.device_list.get("water_heater")
+    entities = []
+    if devices:
+        for dev in devices:
+            entities.append(OSOEnergyWaterHeater(osoenergy, dev))
+    async_add_entities(entities, True)
+
+    platform = entity_platform.async_get_current_platform()
+
+    platform.async_register_entity_service(
+        SERVICE_TURN_ON,
```

Comment:
The water heater entity already has a turn-on service. Why do you need a custom one?